calcom/cal.com
# feat: add MercadoPago payment app

## What does this PR do?



1. Adds MercadoPago as a new payment app for Cal.com,
2. Adds a webhook endpoint to listen for payments created with MercadoPago.

#### About MercadoPago
Mercado Pago is a leading digital payment platform in Latin America. Currently, it's available in seven countries: Argentina, Brazil, Chile, Colombia, Mexico, Uruguay, and Peru. It's widely used by businesses and consumers in all the above countries, and it surpassed $20 billion payment volume in 2019.



Checkout the [Loom Video Demo](https://www.loom.com/share/5746c7952fb444a9aec3caeb55f7191c?sid=0d9b74a5-d376-4be5-b750-b64d4bc168f0) for this PR.

## Requirement/Documentation



- If there is a requirement document, please, share it here.
 No
- If there is ab UI/UX design document, please, share it here.
 No

## Type of change



## How should this be tested?



- Are there environment variables that should be set?
-- Yes, we need the following environment variables: `MERCADOPAGO_CLIENT_ID`, `MERCADOPAGO_CLIENT_SECRET`,  `NEXT_PUBLIC_MERCADOPAGO_PUBLIC_KEY`. See the guide after this bullet points to learn how to retrieve the configs from the App in MercadoPago to Cal.com enviornment variables.
- What are the minimal test data to have?
-- _MercadoPago requirements_: 
--- 1. Signup to MercadoPago [here](https://www.mercadopago.com).
--- 2. Create MercadoPago application in the developers portal [here](https://www.mercadopago.com.pe/developers/panel/app).
--- 3. Retrieve the Client ID on the applications page.
--- 4. Retrieve the Secret on the test credentials section (sidebar).
--- 5. Retrieve the Public Key on the test credentials section (sidebar)
--- 6. Set the Webhook (sidebar). Should point to `YOUR_URL/api/integrations/mercadopago/webhook` and should select "Pagos" (Payments) as the only event needed.
--- 7. Set the Redirect Url on the the application page. Should point to `YOUR_URL/api/integrations/mercadopago/callback`
See the guide after this bullet points to learn how to setup an App in MercadoPago.
-- _Cal.com requirements:_
--- 1. Install MercadoPago payment app.
--- 2. Add the MercadoPago payment information in an EventType
- What is expected (happy path) to have (input and output)?
-- Being able to install MercadoPago payment app in Cal.com.
-- Being able to book and pay an event with MercadoPago in Cal.com.
-- Being able to cancel an event and refund with MercadoPago in Cal.com.
- Any other important info that could help to test that PR.
-- See the following guide.

In order to test the MercadoPago integration:
1. Create a MercadoPago account [here](https://www.mercadopago.com/developers/panel/app). 
2. Go to the [developers site](https://www.mercadopago.com/developers/panel/app) and set up an application. Here you should add the relevant environment variables to the `.env` file. You'll need the Application ID of MercadoPago (`MERCADOPAGO_CLIENT_ID`), the Public Key of MercadoPago (`NEXT_PUBLIC_MERCADOPAGO_PUBLIC_KEY`), the Access Token of MercadoPago (`MERCADOPAGO_CLIENT_SECRET`). After that you will to set up the Webhook URL to point to the `https://yesyesyes.ngrok.app/api/integrations/mercadopago/webhook`. Finally, you should edit the MercadoPago application and add the RedirectURL to point to `https://yesyesyes.ngrok.app/api/integrations/mercadopago/callback`. I've made a [Loom Video](https://www.loom.com/share/9b7d8a04a476496b9b2b6fd0cdf508fb?sid=59614e50-4455-451e-b342-652103592745) to quickly explain this process.
3. Once you had set up the configs in the `.env` file you should run `yarn dx` so that it can seed again the app credentials to the database.
4. Install MercadoPago payment app to your Cal.com account.
5. Create an EventType and add the MercadoPago app. Setup the price and currency. The currency should be the same local currency (or USD) as the business set up in MercadoPago. Since there is no way to retrieve this from the API, the user should manually set the currency.
6. In an incognito window, schedule an Event with the EventType that has the MercadoPago integration. 
7. After confirming the booking you will see the Payment page with the MercadoPago payment button. The currency and price should be reflected here.
9. The pay button will redirect you to MercadoPago account so that you can proceed with the payment. You can used the [test credit cards](https://www.mercadopago.com.pe/developers/es/docs/checkout-pro/additional-content/your-integrations/test/cards) and enter "12345678" for the ID number.
10. After the payment is done you should be redirected to the Booking page, which will indicate that the payment was made.
11. You can also cancel the event and will see that the Payment has been refunded by MercadoPago.

## Mandatory Tasks


## Checklist


- I haven't added tests that prove my fix is effective or that my feature works
